doc/Container Hardening/SONiC_container_hardening_HLD.md

```diff
@@ -0,0 +1,349 @@
+## List of Tables
+* [Table 1: Revision](#table-1-revision)
+* [Table 2: Abbreviations](#table-2-abbreviations)
+* [Table 3: Default Linux capabilities](#table-3-default-linux-capabilities)
+* [Table 4: Extended Linux capabilities](#table-4-extended-linux-capabilities)
+
+## Revision
+###### Table 1: Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 |             |                    | Initial version                   |
+
+## Scope
+
+This section describes the requirements, goals, and recommendations of the container hardening item for SONiC.
+
+## Definitions/Abbreviations
+###### Table 2: Abbreviations
+| Definitions/Abbreviation | Description                                |
+|--------------------------|--------------------------------------------|
+| OS                       | Operating System                           |
+| API                      | Application Programmable Interface         |
+| SAI                      | Swich Abstraction Interface                |
+
+## 1. Overview
+
+Containers is a method of creating virtualization and abstraction of an OS for a subset of processes/service on top of a single host with the purpose of giving it an environment to run and execute its tasks without effect of nearby containers/processes.
+
+In SONiC, we are deploying containers with full visibility and capabilities as the host Linux.
+
+This poses a security risk and vulnerability as a single breached container means that the whole system is breached.
+
+Addressing this issue – we have composed this doc for container hardening, describing the security hardening requirements and definitions for all containers on top of SONiC.
+
+## 2. Requirements
+
+What are we trying to achieve here?
+
+We would like to increase the security in SONiC so that an attack on a specific container will not compromise the whole system.
+
+To do so, we'll tackle the following areas:
+1. Privileges
+2. Network
+3. Capabilities
+4. Mount namespace
+5. Cgroups
+6. Etc
+
+For now, we will focus on #1 & #2
+
+Further guidelines and requirements will be brought upon in the future on-demand.
+
+## 3. Architecture Design
+
+### 3.1 Root privileges
+
+When removing the root privileges from a specific container - we are required to remove the `--privileged` flag and add the required missing Linux capabilities to the docker, or alternatively adjust the container so that it does not require root privileges to perform any action.
+
+### 3.2 net=host
+
+Removing the `net=host` is required to prevent the container from accessing the full network scope of the host and system.
+When doing this removal - we will start getting failures from devices that require external access and packet transfers between the container and the host to the interfaces.
+In order to overcome this obstacle - we have a few options here:
+- using `--net=bridge` and port forwarding
```

Comment:
Is there any plan to harden pmon containers ?

Reply:
List of potential container hardening will be added.